corehq/apps/sso/tests/test_backends.py

```diff
@@ -0,0 +1,225 @@
+from django.contrib import auth
+from django.test import TestCase, RequestFactory
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+from corehq.apps.sso.tests import generator
+from corehq.apps.users.models import WebUser
+
+
+class TestSsoBackend(TestCase):
+    """
+    This ensures that authentication through the SsoBackend only succeeds
+    when a samlSessionIndex is present and the Identity Provider, its
+    associated email domains, and the username all meet required criteria.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.account = generator.get_billing_account_for_idp()
+        cls.domain = Domain(
+            name="vaultwax-001",
+            is_active=True
+        )
+        cls.domain.save()
+
+        # this will be the user that's "logging in" with SAML2 via the SsoBackend
+        cls.user = WebUser.create(
+            cls.domain.name, 'b@vaultwax.com', 'testpwd', None, None
+        )
+        cls.idp = generator.create_idp('vaultwax', cls.account)
+        cls.idp_two = generator.create_idp('vwx-link', cls.account)
+
+    @classmethod
+    def tearDownClass(cls):
+        IdentityProvider.objects.all().delete()
+        cls.user.delete(None)
+        cls.domain.delete()
+        cls.account.delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        super().setUp()
+        # Identity Provider is_active is always True by default for each test
+        self.idp.is_active = True
+        self.idp.save()
```

Comment:
Do tests cause this to be reset to false? If not, could it be moved into `setUpClass`?

Reply:
some tests set this to `False` which is why it's not in `setUpClass`

Reply:
Looks like only one test does that. What do you think of using `addCleanup` in that test? Otherwise it's an extra database hit for every test.

Reply:
i think i might just add this to the one test that sets it to false